keps/sig-apps/2232-jobs-stopped/README.md

```diff
@@ -0,0 +1,546 @@
+	// false to true), the Job controller will delete all active Pods associated
+	// with this Job. Users must design their workload to gracefully handle this.
+	// This is an alpha field and requires enabling the StoppedJob feature gate.
+	// Defaults to false.
+	Stopped *bool `json:"stopped,omitempty"`
+
+	...
+}
+```
+
+As described in the comment, when the boolean is set to true, the
+controller-manager abstains from creating Pods even if the Job's parallelism
+requirements aren't met. If the Job is already active and is updated with
+`stopped: true`, the Job controller calls Delete on all its active Pods. This
+causes the kubelet to send a SIGTERM signal and completely remove the Pod after
+its graceful termination period is honoured. Pods terminated this way are
+considered a failure and the controller does not count terminated Pods towards
+completions. This behaviour is similar to decreasing parallelism to zero.
+
+Similar to existing [JobConditionType](https://github.com/kubernetes/kubernetes/blob/c98f6bf30890f2c5826067ae50cfc36958106e68/staging/src/k8s.io/api/batch/v1/types.go#L167)s
+"Complete" and "Failed", we propose adding a new condition type called
+"Stopped" as a part of the Job's status:
+
+```go
+type JobStatus struct {
+	// The latest available observations of an object's current state. Can be
+	// one of "Stopped", "Failed", or "Complete".
+	Conditions []JobCondition `json:"conditions,omitempty"`
+
+	...
+}
+
+// JobCondition describes current state of a job.
+type JobCondition struct {
+	// Type of job condition, Complete or Failed.
+	Type JobConditionType `json:"type"`
+
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus `json:"status"`
+
+	// Last time the condition transit from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+
+	...
+}
+
+type JobConditionType string
+
+// These are valid conditions of a job.
+const (
+	// JobStopped means the job has been stopped.
+	JobStopped JobConditionType = "Stopped"
+	// JobComplete means the job has completed its execution.
+	JobComplete JobConditionType = "Complete"
+	// JobFailed means the job has failed its execution.
+	JobFailed JobConditionType = "Failed"
+)
+```
+
+To determine if a Job has been stopped, users must look for the JobCondition
+with `Type` as "Stopped" and `Status` as true. If such a JobCondition does not
+exist, the Job has never been stopped. Otherwise, if the `Status` is false, the
+Job was previously stopped and then resumed. Inferring this from the Job spec
+is not recommended as the Job controller may not have seen the update yet. When
+a Job is stopped, the Job controller sets the condition only after all Pods are
+terminating or have been deleted.
+
+The Stopped JobCondition also has a `LastTransitionTime` field. This can be
+used to infer how long a Job has been stopped for (if the `Status` value is
+true).
```

Comment:
out of curiosity, when updating an existing job condition from true to false, does this happen in place or as an addition entry in the Conditions list?

Reply:
Since job is a finite resource, we only add new conditions. Currently, only https://github.com/kubernetes/kubernetes/blob/9f2f1b8c460b4b5ec8bfcd34490fb259ebad65b0/pkg/controller/job/job_controller.go#L534 and https://github.com/kubernetes/kubernetes/blob/9f2f1b8c460b4b5ec8bfcd34490fb259ebad65b0/pkg/controller/job/job_controller.go#L568

Reply:
Wait - are you saying that a Job may have multiple conditions with the same type? That sounds wrong...

Reply:
I am guessing currently no since the two conditions we now have (Failed and Completed) are terminal once the job reaches to either.

This KEP proposes adding a new condition that we can exit from (Stopped). I am not sure what is the common pattern when dealing with such conditions, do we append a new entry or update the existing one? I am guessing the former.

Reply:
We update the existing one.

Reply:
In other words: for a given condition type, there should always be at most one entry.

Reply:
sg, I updated the text to emphasize that.